cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -1374,10 +1139,35 @@ Status GetListCastFunc(const DataType& in_type, const std::shared_ptr<DataType>&
 
 }  // namespace
 
+inline bool IsZeroCopyCast(Type::type in_type, Type::type out_type) {
+  switch (in_type) {
+    case Type::INT32:
+      return (out_type == Type::DATE32) || (out_type == Type::TIME32);
+    case Type::INT64:
+      return ((out_type == Type::DATE64) || (out_type == Type::TIME64) ||
+              (out_type == Type::TIMESTAMP));
+    case Type::DATE32:
+    case Type::TIME32:
+      return out_type == Type::INT32;
+    case Type::DATE64:
+    case Type::TIME64:
+    case Type::TIMESTAMP:
+      return out_type == Type::INT64;
+    default:
+      break;
+  }
+  return false;
+}
```

Comment:
This was being maintained with templates before... that was overly elaborate. If this gets too complicated, we can easily code-generate it